### PR TITLE
Upgrades the resolver version to lts-19.6 ghc to 9.0.2.

### PR DESCRIPTION
**Pull request by rimbi**
## Purpose

Update Haskell toolchain and dependencies.

## Changes

- Upgrade resolver to lts-19.6
- Upgrade compiler to 9.0.2
- Apply the changes in the dependent APIs.
- Add an installer for the macOS distribution
   - This was necessary because we had to use dynamic linking to avoid a GHC issue that only appeared when statically linking the client on intel macs.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: stack.yaml [hunk 1]
```diff
@@ -78,7 +77,9 @@ extra-lib-dirs:
 #
 # Allow a newer minor version of GHC than the snapshot specifies
 # compiler-check: newer-minor
-# allow-newer: true
+allow-newer: true
+extra-include-dirs:
+- /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include/ffi/
```
Comment (abizjak) [inline stack.yaml:82]: Same comment as in the other PRs. This cannot remain.

**Review by abizjak — COMMENTED** [review]
I think this looks good, apart from the extra-include-dirs directive.

File: README.md [hunk 2]
```diff
@@ -2,28 +2,35 @@
...
-* deploy, initialize, and interact with smart contracts,
-* view and manage local configurations such as aliases for accounts, modules and contracts,
-* query data from the chain,
-* query the state of the consensus protocol, and
-* inspect and manage the node.
+- deploy, initialize, and interact with smart contracts,
+- view and manage local configurations such as aliases for accounts, modules and contracts,
+- query data from the chain,
+- query the state of the consensus protocol, and
+- inspect and manage the node.
```
Comment (abizjak) [inline README.md:33]: This is really not necessary and is just change.

File: README.md [hunk 3]
```diff
@@ -35,52 +42,79 @@ Windows](https://developer.concordium.software/en/mainnet/net/installation/downl
...
+  - Via [GHCup](https://www.haskell.org/ghcup/):
+    - Unix: `curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | sh`
+    - Windows: [Follow the GHCup install guide](https://www.haskell.org/ghcup/install/#installation).
+  - Or, by itself:
+    - Unix: `curl -sSL https://get.haskellstack.org/ | sh`
+    - Windows: [Follow the Stack install guide](https://docs.haskellstack.org/en/stable/install_and_upgrade/).
+
+- Install Rust version 1.53+:
+  - Unix: `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh`
+  - Windows: [Follow the Rust install
```
Comment (abizjak) [inline README.md:55]: And this. There was nothing wrong with what was there.

File: README.md [hunk 4]
```diff
@@ -35,52 +42,79 @@ Windows](https://developer.concordium.software/en/mainnet/net/installation/downl
...
-* `BACKEND` is the GRPC server on which to perform the actions/queries.
+- `BACKEND` is the GRPC server on which to perform the actions/queries.
   It's specified using the flags `--grpc-ip`, `--grpc-port`, and `--grpc-target`
   (might be needed when calling through proxy like, say, on the testnet).
 
-* `COMMAND` is a command from one of the categories described below in [Commands](#commands).
+- `COMMAND` is a command from one of the categories described below in [Commands](#commands).
 
-* `ARGS` is the list of arguments provided to `COMMAND`.
+- `ARGS` is the list of arguments provided to `COMMAND`.
```
Comment (abizjak) [inline README.md:117]: None of this was necessary.

File: README.md [hunk 5]
```diff
@@ -163,12 +199,12 @@ The variable `XDG_CONFIG_HOME` is defined by the
 as the location of user specific configuration.
 If not set or empty, it has the following system-dependent defaults:
 
-* Unix: `$HOME/.config`
-* Windows: `%APPDATA%` (`C:\Users\<user>\AppData\Roaming`)
+- Unix: `$HOME/.config`
+- Windows: `%APPDATA%` (`C:\Users\<user>\AppData\Roaming`)
```
Comment (abizjak) [inline README.md:203]: Nor this.

File: stack.yaml [hunk 6]
```diff
@@ -78,7 +77,7 @@ extra-lib-dirs:
 #
 # Allow a newer minor version of GHC than the snapshot specifies
 # compiler-check: newer-minor
-# allow-newer: true
+allow-newer: true
```
Comment (abizjak) [inline stack.yaml:80]: Why? This should not be enabled unless absolutely necessary. We want the dependency versions from the snapshot.

**Review by abizjak — COMMENTED** [review]
Can you link to the test builds of the client? As part of the review I would like to verify that they are as they should be.

File: README.md (hunk 3 above)
Comment (rimbi) [reply README.md:55]: It's about consistency, something markdown linter complains about.

File: README.md (hunk 4 above)
Comment (rimbi) [reply README.md:117]: It's about consistency, something markdown linter complains about.

File: README.md (hunk 5 above)
Comment (rimbi) [reply README.md:203]: It's about consistency, something markdown linter complains about.

File: stack.yaml (hunk 6 above)
Comment (rimbi) [reply stack.yaml]: I have enabled it after a build error, checking again...

**Comment by rimbi** [comment]
> Can you link to the test builds of the client? As part of the review I would like to verify that they are as they should be.

What do you mean exactly?

File: stack.yaml (hunk 6 above)
Comment (rimbi) [reply stack.yaml]: done

File: concordium-client.cabal [hunk 7]
```diff
@@ -37,7 +37,7 @@ flag middleware
 flag static
   description: Use static linking.
   manual: True
-  default: False
+  default: True
```
Comment (abizjak) [inline concordium-client.cabal:40]: I don't think this should have been changed. Static linking is troublesome during development. Can you change it back to False as the default please.

File: ChangeLog.md [hunk 8]
```diff
@@ -3,13 +3,16 @@
 ## Unreleased
 
 ## 4.0.3
+
 - Fix calculation of payload size of a configure baker transaction in
   the function `bakerConfigurePayloadSize` in Transaction.hs used by
   the wallet-proxy.
 - Fix help text of `baker add` and `baker configure` commands that referred to
   `raw GetChainParameters`.
+- Upgrade resolver to version `lts-19.6` and ghc to version `9.0.2`.
```
[the remark this answers is not in the document]
Comment (rimbi) [reply ChangeLog.md]: done

File: concordium-client.cabal (hunk 7 above)
Comment (rimbi) [reply concordium-client.cabal]: done

**Review by abizjak — APPROVED** [review]

**Pull request merged**